sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala

```diff
@@ -21,6 +21,15 @@ import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.expressions._
 
+// for lazy computing, be sure the generator.terminate() called in the very last
+private[execution] sealed case class LazyIterator(func: () => TraversableOnce[Row])
+  extends Iterator[Row] {
+
+  lazy val results = func().toIterator
+  override def hasNext: Boolean = results.hasNext
+  override def next(): Row = results.next()
+}
```

Comment:
@liancheng I just confirmed, the `CompletionIterator` provides a callback machinism once finish iterating the iterator, however, all we need here is to append an new iterator, which produced by the `terminate()` method.

But you're right, probably we need to update the `CompletionIterator` a little bit, I leave a `TODO` here, and will file another PR for this improvement.